packages/@expo/cli/src/verifyNativeAssets/verifyNativeAssets.ts

```diff
@@ -0,0 +1,172 @@
+import fs from 'fs';
+import glob from 'glob';
+import path from 'path';
+
+import {
+  BuildManifest,
+  ExportedMetadata,
+  ExportedMetadataAsset,
+  FullAssetDump,
+  FullAssetDumpEntry,
+  Platform,
+} from './types';
+import { CommandError } from '../utils/errors';
+
+const debug = require('debug')('expo:verify-native-assets') as typeof console.log;
+
+/**
+ * Finds any assets that will be missing from an app given a build and an exported update bundle.
+ *
+ * @param buildPath Path to an EAS or local build containing an expo-updates embedded manifest (app.manifest)
+ * @param exportPath Path to a directory produced by the command `npx expo export --dump-assetmap`
+ * @param platform Either `android` or `ios`
+ * @param projectRoot The project root path
+ * @returns An array containing any assets that are found in the Metro asset dump, but not included in either app.manifest or the exported bundle
+ */
+export function getMissingAssets(
+  buildPath: string,
+  exportPath: string,
+  platform: Platform,
+  projectRoot: string
+) {
+  const buildManifestHashSet = getBuildManifestHashSet(
+    getBuildManifest(buildPath, platform, projectRoot)
+  );
+
+  const fullAssetMap = getFullAssetDump(exportPath);
+  const fullAssetSet = getFullAssetDumpHashSet(getFullAssetDump(exportPath));
+
+  const exportedAssetSet = getExportedMetadataHashSet(getExportedMetadata(exportPath), platform);
+
+  debug(`Assets in build: ${JSON.stringify([...buildManifestHashSet], null, 2)}`);
+  debug(`Assets in exported bundle: ${JSON.stringify([...exportedAssetSet], null, 2)}`);
+  debug(`All assets resolved by Metro: ${JSON.stringify([...fullAssetSet], null, 2)}`);
+
+  const buildAssetsPlusExportedAssets = new Set(buildManifestHashSet);
+  exportedAssetSet.forEach((hash) => buildAssetsPlusExportedAssets.add(hash));
+
+  const missingAssets: FullAssetDumpEntry[] = [];
+
+  fullAssetSet.forEach((hash) => {
+    if (!buildAssetsPlusExportedAssets.has(hash)) {
+      const asset = fullAssetMap.get(hash);
+      console.warn(`  Missing asset: hash = ${hash}, file = ${asset?.files[0] ?? ''}`);
+      asset && missingAssets.push(asset);
+    }
+  });
+
+  return missingAssets;
+}
+
+/**
+ * Reads and returns the embedded manifest (app.manifest) for a build.
+ *
+ * @param buildPath Path to the build folder
+ * @param platform Either 'android' or 'ios'
+ * @param projectRoot The project root path
+ * @returns the JSON structure of the manifest
+ */
+export function getBuildManifest(buildPath: string, platform: Platform, projectRoot: string) {
+  let realBuildPath = buildPath;
+  if (buildPath === projectRoot) {
+    switch (platform) {
+      case 'android':
+        realBuildPath = path.resolve(projectRoot, 'android', 'app', 'build');
```

Comment:
> Path to an EAS or local build containing an expo-updates embedded manifest (app.manifest)

At first I was under the impression that one could pass in the final build apk, aab, or ipa, when in reality they need the build artifacts/intermediates. I'm not sure how reliable this method (using build intermediates) will be for determining compatibility for a few reasons:
- artifacts are deleted/cleaned/removed often, and sometimes even automatically. the final build is often the only output that is kept.
- locally, it is overwritten for each build and often stale artifacts/intermediates remain in the unclean'ed directory from prior builds.
- users won't have these for an EAS build.

So, I guess my question is about proposed usage of this. Are we requiring users to retain copies of these intermediates? Or, should we provide a way to decompile the binary to extract the app.manifest? Or, should we just record this info when making an EAS build and provide that info in a raw form into this command?

Reply:
@wschurman the apk or ipa would be fine, since they would presumably have the `app.manifest` file. They would have to be unpacked first.

Reply:
There is already an `apktool` to unpack Android apks, and I think ipas are just zip or gzip files.

Reply:
Yep, I think we just need to be deliberate about what the input to this function should be. IMO it makes the most sense to pass the app.manifest file path explicitly so that this command isn't responsible for digging in IPA/AAR/APK/ZIP files or build intermediates. It also leaves the window open for supplying the app.manifest directly from EAS servers, which would potentially be a good UX at update time.

Reply:
@wschurman now that I've moved this to the expo-updates CLI, it might be good to be able to pass the app.manifest directly with a different command option. I'll try adding that next.

Reply:
Sounds good. My take is that this command should be as single-responsibility as possible and just take the two file paths as arguments (the assetmap and the app.manifest). The single responsibility is determining what the missing assets are as opposed to that plus knowing the structure of builds/dist folder. My guess is that this will most often be run by us from eas-cli or something, and that CLI knows exactly where these two files live (it does the export and also does the build and can fetch build metadata/app.manifest for a build from the servers).